docs/BLOCK_DEVICE_LOCKING.md

```diff
@@ -98,3 +98,145 @@ Summarizing: it is recommended to take `LOCK_EX` BSD file locks when
 manipulating block devices in all tools that change file system block devices
 (`mkfs`, `fsck`, …) or partition tables (`fdisk`, `parted`, …), right after
 opening the node.
+
+# Sample of Locking The Whole Disk
+
+The following is a sample to leverage `libsystemd` infrastructure to get the whole disk of a block device and add BSD lock on it.
+
+## Compile and Execute
+Note the version of `libsystemd` should be equal or greater than v251.
+
+Place the code in a source file, e.g. `main.c`. Assuming the version of systemd is satisfied, run the following commands:
+```
+# gcc -o main -lsystemd main.c
+
+# ./main /dev/sda1
+Successfully get the lock: /dev/sda
+
+# flock -x /dev/sda ./main /dev/sda1
+Failed to get the lock: /dev/sda
+Error happened: -1
+```
+
+If the version of `systemd` is not satisfied, download and compile the satisfied version of `systemd` source code. Assuming the code of `systemd` is placed in `/path/to/systemd/` and the targets are placed in `/path/to/systemd/build/`, run the following commands:
+```
+# gcc main.c -L/path/to/systemd/build/ -lsystemd -I/path/to/systemd/src/ -o main
+
+# LD_LIBRARY_PATH=/path/to/systemd/build:${LD_LIBRARY_PATH} ./main /dev/sda1
+Successfully get the lock: /dev/sda
+
+# LD_LIBRARY_PATH=/path/to/systemd/build:${LD_LIBRARY_PATH} flock -x /dev/sda ./main /dev/sda1
+Failed to get the lock: /dev/sda
+Error happened: -1
+```
+
+## Code
+```
+/* SPDX-License-Identifier: MIT-0 */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/file.h>
+#include <systemd/sd-device.h>
+#include <unistd.h>
+
+/**
+ * lock_whole_disk_from_devname
+ * @devname: devname of a block device, e.g., /dev/sda or /dev/sda1
+ * @open_mode: the mode to open the device, e.g., O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY
+ * @lock_mode: the mode to add flock, e.g., LOCK_EX|LOCK_NB
+ * @ret_fd: return the file description if successfully locked
```

Comment:
"description" → "descriptor"

Why doesn't the function return fd via the return value?

Reply:
I just want to return the error number. Is there any elegant alternatives, e.g., overlay the `errno`?

Reply:
Just use the usual systemd convention: return negative errno value on error, or non-negative fd.